lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,259 @@
+	client := resty.NewWithClient(defaults.Config().HTTPClient)
+	client.SetHeader("Authorization", "GenieKey "+conf.APIKey)
+	client.SetHostURL(conf.APIEndpoint)
+	return &Client{
+		client:       client,
+		ClientConfig: conf,
+	}, nil
+}
+
+func errWrapper(statusCode int, err error) error {
+	switch statusCode {
+	case http.StatusForbidden:
+		return trace.AccessDenied("opsgenie API access denied: %v", err)
+	case http.StatusRequestTimeout:
+		return trace.ConnectionProblem(err, "connecting to opsgenie API")
+	}
+	return trace.Wrap(err)
+}
+
+// CreateAlert creates an opsgenie alert.
+func (og Client) CreateAlert(ctx context.Context, reqID string, reqData RequestData) (OpsgenieData, error) {
+	bodyDetails, err := buildAlertBody(og.WebProxyURL, reqID, reqData)
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	body := AlertBody{
+		Message:     fmt.Sprintf("Access request from %s", reqData.User),
+		Alias:       fmt.Sprintf("%s/%s", alertKeyPrefix, reqID),
+		Description: bodyDetails,
+		Responders:  og.getResponders(reqData),
+		Priority:    og.Priority,
+	}
+
+	var result AlertResult
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetResult(&result).
+		Post("alerts"); err != nil {
+		return OpsgenieData{}, errWrapper(resp.StatusCode(), err)
+	}
+
+	return OpsgenieData{
+		AlertID: result.Alert.ID,
+	}, nil
+}
+
+func (og Client) getResponders(reqData RequestData) []Responder {
+	schedules := og.DefaultSchedules
+	if reqSchedules, ok := reqData.RequestAnnotations[ReqAnnotationRespondersKey]; ok {
+		schedules = reqSchedules
+	}
+	responders := make([]Responder, 0, len(schedules))
+	for _, s := range schedules {
+		responders = append(responders, Responder{
+			Type: "schedule",
+			ID:   s,
+		})
+	}
+	return responders
+}
+
+// PostReviewNote posts a note once a new request review appears.
+func (og Client) PostReviewNote(ctx context.Context, alertID string, review types.AccessReview) error {
+	note, err := buildReviewNoteBody(review)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/notes"); err != nil {
+		return errWrapper(resp.StatusCode(), err)
+	}
+	return nil
+}
+
+// ResolveAlert resolves an alert and posts a note with resolution details.
+func (og Client) ResolveAlert(ctx context.Context, alertID string, resolution Resolution) error {
+	note, err := buildResolutionNoteBody(resolution)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/close"); err != nil {
```

Comment:
I think we want to empty and close `resp.Body` after the call if err!= nil. Else the HTTP client cannot reuse connections and we might leak resources.

According to [http.Client docs](https://pkg.go.dev/net/http#Client.Do)
> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.